### PR TITLE
Delay data provider instantiation to avoid caching too many tiles when using cbcls.

### PR DESCRIPTION
**Pull request by jacarey**
Since the data provider for cbcls caches an entire tile of data, and we weren't delaying its creation prior to running extraction, `ExtractIlluminaBarcodes` was running out of memory.

This PR delays instantiation. In addition it reads only header data until `advance()` is called and data is requested.  This limits the number of tiles that can be cached to the total number of barcode extraction threads.

### Checklist (never delete this)

Never delete this, it is our record that procedure was followed. If you find that for whatever reason one of the checklist points doesn't apply to your PR, you can leave it unchecked but please add an explanation below.

#### Content
- [x] Added or modified tests to cover changes and any new functionality
- [x] Edited the README / documentation (if applicable)
- [ ] All tests passing on Travis

#### Review
- [ ] Final thumbs-up from reviewer
- [ ] Rebase, squash and reword as applicable

For more detailed guidelines, see https://github.com/broadinstitute/picard/wiki/Guidelines-for-pull-requests

### CODE REVIEW

**Comment by jacarey** [comment]
@gbggrant Found this bug while testing an S4 novaseq. This is a bug in the refactor that we pushed a few months ago.

**Comment by yfarjoun** [comment]
Sorry @gbggrant, I am not familiar with this part of the code and I do not have the time to ramp up on it... :-(

**Review by yfarjoun — COMMENTED** [review]
I'm not familiar with this part of the code and cannot bring myself up to speed on it.

File: src/main/java/picard/illumina/parser/readers/CbclReader.java [hunk 1]
```diff
@@ -268,6 +265,9 @@ public void close() {
     }
 
     private void advance() {
+        if(!tileCached){
```
Comment (gbggrant) [inline src/main/java/picard/illumina/parser/readers/CbclReader.java:268]: spacing:
```suggestion
        if (!tileCached) {
```

File: src/main/java/picard/illumina/parser/readers/CbclReader.java [hunk 2]
```diff
@@ -187,39 +186,37 @@ private void readSurfaceTile(final int tileNum, final List<AbstractIlluminaPosit
...
-                    } catch (final IOException e) {
-                        // when logging the error, increment cycle by 1, since totalCycleCount is zero-indexed but Illumina directories are 1-indexed.
-                        throw new PicardException(String.format("Error while reading from BCL file for cycle %d. Offending file on disk is %s",
-                                (totalCycleCount + 1), this.streamFiles[totalCycleCount].getAbsolutePath()), e);
+                        cacheTile(totalCycleCount, cycleData[totalCycleCount].tileInfo, currentCycleData);
                     }
-                    totalCycleCount++;
+                } catch (final IOException e) {
+                    // when logging the error, increment cycle by 1, since totalCycleCount is zero-indexed but Illumina directories are 1-indexed.
+                    throw new PicardException(String.format("Error while reading from BCL file for cycle %d. Offending file on disk is %s",
```
Comment (gbggrant) [inline src/main/java/picard/illumina/parser/readers/CbclReader.java:213]: Should this be CBCL?
```suggestion
                    throw new PicardException(String.format("Error while reading from CBCL file for cycle %d. Offending file on disk is %s",
```

**Review by gbggrant — APPROVED** [review]
@jacarey looks good.  Two super minor nits.  I will go ahead and merge this in the near future if I don't hear from you on them.  Sorry for the delay in getting to this.

**Pull request merged**